man/hdp.1
.\" $Id$
.TH HDP 1 "July 1995" "NCSA HDF 4.0"

.nf

NAME
     hdp - HDF dumper

SYNOPSIS

DESCRIPTION
     
     hdp is a command line utility designed for quick display of 
     contents and data of HDF objects. It can list the contents 
     of hdf files at various levels with different details. It can 
     also dump the data of one or more specific objects in the file. 


HDP OPTIONS

    Currently, there is only one option.

    -H  Display usage information about the specified command.
        If no command is specified, -H lists all available commands.


HDP COMMANDS

     hdp currently has two types of commands: list and dump. Other 
     types of commands such as those for editing may be added in the
     future.
     
     hdp list <filename list>
         lists contents of files in <filename list> 

     hdp dumpsds <filename list>
         displays data of NDGs and SDGs in the listed files.

     hdp dumpvd <filename list>
         displays data of vdatas in the listed files.
   
     hdp dumpvg <filename list>
         displays data of objects in vgroups in the listed files.

     hdp dumprig <filename list>
         displays data of RIGs in the listed files.


HDP COMMAND OPTIONS

     (note: options preceeded by an * have not yet been implemented.)


.ul 2
     hdp list [format options] [content ops] [filter ops] [order ops]
        <filename list>

      Format options
          decide how the info of objects will be presented on the screen.
       
      -s  (short format) under each tag #, all ref's of that tag are listed
          in one or more lines, same as the output of hdfls. (default)

      -l  (long format) one object per line. Each line contains tag-name, 
          tag/ref and the index of this tag in the file.(e.g., the ith NDG 
          in the file).

      -d  debug format, one object per line. Each line contains tag_name,
          tag/ref, index, offset, and length, same as the output of hdfls -d.

.ul
      no      tagname    tag    ref   index/tag       offset  length

      1       DFTAG_NT   106      2      1          
      2       DFTAG_SD   701      3      1
      ...


       Content options
            allow contents be displayed.

          -n  display the name or label of the object, if there is any.
              -n puts you in -l format automatically.

          -c  display the class of the object, if there is any. -l format.

          -a  display description of the object, if there is any. -l format.

       Filter options
            select certain type of objects to display, default is all.

          -g  display groups only. Objects which do not belong to 
              any group will not be displayed. Nested groups will be
              displayed in tree format.

          -t <number>  display objects with specified tag number . e.g. 
                       720 for NDG.
          -t <name>    display objects with specified tag name.

       Order options
            sort the output list in different orders.
 
          -ot  by tag # (default)
          -of  by the order in file DDlist.
          -og  by group
          -on  by name(label)

.ul
    hdp dumpsds [filter ops] [contents ops] [output ops] <filename list>
         Filter options
              specify which SDS to dump.

             -i <index>   dump SDS's with indices specified in <index>; indices
                          correspond to the order of the SDS in the file
             -r <ref>     dump SDS's with reference numbers specified in <ref>
             -n <name>    dump SDS's with names specified in <name>
             -a           dump all SDSs in the file. (default)

          Content options

             -v    display everything including all annotations (default)
             -h    header only, no annotation for elements or data
             -d    data only, no tag/ref 

          Output options

             -o \fIfn\fP      specify \fIfn\fP as output file name
           * -b         binary file
             -t         ascii text file (default)

          Note: any combination of an option from each of the three categories
                can be used; but no more than one option from one category is 
                allowed.

.ul
    hdp dumpvd [filter ops] [contents ops] [output ops] <filename list>
         Filter options
              specify which vdata to dump.

             -i <index>   dump vdatas with indices in <index>; indices 
                          correspond to the order of the vdatas in the 
                          files
             -r <ref>     dump vdatas with reference numbers specified in
                          <ref>
             -n <name>    dump vdatas with names specified in <name>
             -c <class>   dump vdatas with classes specified in <class>
             -a           dump all vdatas in the file. (default)

          Content options

             -v    display everything including all annotations (default)
             -h    header only, no annotation for elements or data
             -d    data only, no tag/ref
             -f <fields> dump data of specified fields

          Output options

             -o \fIfn\fP      specify \fIfn\fP as output file name
           * -b         binary file
             -t         text ascii file (default)

.ul
    hdp dumpvg [filter ops] [contents ops] [output ops] <filename list>
         Filter options
              specify which vgroups to dump.

             -i <index>   dump vgroups with indices specified in <index>; 
                          indices correspond to the order of the vgroups 
                          specified in the files
             -r <ref>     dump vgroups with reference numbers specified in <ref>
             -n <name>    dump vgroups with names specified in <name>
             -c <class>   dump vgroups with classes specified in <class>
             -a           dump all vgroups in the file. (default)

          Content options

             -v    display everything including all annotations (default)
             -h    header only, no annotation for elements or data
             -d    data only

          Output options

             -o \fIfn\fP      specify \fIfn\fP as output file name
           * -b         binary file
             -t         text ascii file (default)

    Note: Unless the "-d" option is specified, a graphical representation of
          the file will be given after the data has been displayed. 

.ul
    hdp dumprig [filter ops] [contents ops] [output ops] <filename list>
         Filter options
              specify which RIG to dump.

             -i <index>   dump RIGs with indices specified in <index>; 
                          indices correspond to the order of the RIGs 
                          specified in the files.
             -r <ref>     dump RIGs with reference numbers specified in <ref>
             -a           dump all RIGs in the file. (default)
             -m 8 | 24    dump the RIGs of 8-bit or 24-bit

          Content options

             -v    display everything including all annotations (default)
             -h    header only, no annotation for elements or data
             -d    data only

          Output options

             -o \fIfn\fP      specify \fIfn\fP as output file name
           * -b         binary file
             -t         text ascii file (default)




.fi


mfhdf/CHANGES
$Id$

The default fill values for double and floating-point data were changed.
The previous strategy of using numbers that mapped to IEEE NaN under XDR
caused portability problems.  Now FILL_FLOAT and FILL_DOUBLE are large
constatns within the floating-point range.  As before, we continue to
recommend that users define their own variable-specific fill values rather
than use the default ones provided.

The way in which the default fill values are defined in the FORTRAN
interface was changed, so that now the FORTRAN default fill values will
always be the same as the C default values.  This eliminates the need for
the platform-specific foo0.inc files where these constants were previously
defined.

Support for generalized hyperslabs was added.  This adds the new functions
ncvarputg and ncvargetg to the C interface, and new subroutines NCVPTG,
NCVPGC, NCVGTG, and NCVGGC to the Fortran interface.

The FILL parameters defined in the netcdf.inc include file for integer types
for the FORTRAN interface were "fixed" to be the same as the FILL parameters
in the C interface.  In previous releases these had differed from the C
interface values, which meant C and FORTRAN programs would not agree on
default fill values.  This fix is unfortunate but necessary to bring the two
interfaces back into line.

The installation procedure on UNIX platforms was changed to use the GNU
configure approach to constructing makefiles based on inquiries run from a
configure shell script.

Names acceptable to ncgen for
	netcdf name { ... }
may now start with a non-alphabetic character and include any characters
except blanks or "{" characters.

Some new informational files were added at the top-level:  FAQ for answers
to frequently asked questions, ORIGIN to describe where the distribution
came from, COPYRIGHT separates out the Copyright notice from the README
file.

Changed recommended netCDF file extension from ".cdf" to ".nc", and added
"-b" option of ncgen to produce ".nc" files.

Added several options to ncdump to control annotations in output.

An experimental draft prototype C++ interface has been added.  See the c++
subdirectory for the implementation, an example of the use, and soime
preliminary minimal documentation.

Many "const" declarations were added to the interfaces and documentation, to
specify where functions do not change values through passed in pointers.

Added three new C convenience functions to the interface to handle record
I/O.  ncrecget, ncrecput, and ncrecinq permit accessing a whole record's
worth of data with a single call.

Made clearer in the User's Guide that a NULL pointer can be provided for any
return parameter from an inquire function, to indicate that you don't want
that value returned.  Also indicated that inquire functions never incur any
I/O.


------ Mon Dec 23 MST 1991  Changes in netCDF (2.01 to 2.02)

A "memory leak" is fixed that would cause any program that opened and closed
many netCDF files to grow.

All inquire functions (ncinquire, ncdiminq, ncvarinq, ncattinq) now permit a
NULL pointer to be given for any of their return parameters, to indicate
when no return value for that parameter is desired.

The default library type that is built by the distributed Makefile is now
"nonsharable" rather than "sharable".

The format that ncdump uses to output byte data is changed from octal to
decimal.

------ Mon Nov 25 1991	Changes in netCDF (2.0 to 2.01)

Correct usage of the XDR_LD macro is added to all appropriate
program-building rules in all appropriate sub-Makefiles.  This is needed for
proper program-creation under the IRIX operating-system.

Corrects the major.minor revision ID for SunOS sharable libraries to 2.01.

Adds support for the creation of the netCDF Fortran interface under the IRIX
operating-system.

------ Wed Oct  9 1991	Changes in netCDF (1.11 to 2.0)

The netCDF C interfaces were changed to use long values rather than ints for
netCDF dimensions on all machines.  This changed the following interfaces that
formerly use "int" or "int*" for dimensions and dimension pointers:

  int ncdimdef(int cdfid, const char* name, int length);
  int ncdiminq(int cdfid, int dimid, char* name, int* length);
  int ncvarput1(int cdfid, int varid, int *coords, void* value);
  int ncvarget1(int cdfid, int varid, int *coords, void* value);
  int ncvarput(int cdfid, int varid, int *start, int *count, void* value);
  int ncvarget(int cdfid, int varid, int *start, int *count, void* value);

to use "long" or "long*" instead:

  int ncdimdef(int cdfid, const char* name, long length);
  int ncdiminq(int cdfid, int dimid, char* name, long* length);
  int ncvarput1(int cdfid, int varid, long *coords, void* value);
  int ncvarget1(int cdfid, int varid, long *coords, void* value);
  int ncvarput(int cdfid, int varid, long *start, long *count, void* value);
  int ncvarget(int cdfid, int varid, long *start, long *count, void* value);

No change to data representation or file formats was required, since the XDR
form of integers is 32 bits.  The main advantage of this change is that on
platforms such as MSDOS machines on which integers are only 16 bits, it is
now possible to access data from variables that have a dimension greater
than 32767.  This is of some practical importance, since CD-ROM disks are
being written with large datasets in netCDF form, and it is desirable to be
able to access this data from inexpensive 16-bit machines.

Since ints and longs are the same on most current platforms (Suns,
DECstations, VAXen, IBM RS6000s, SGIs, HPs, CRAYs, ...), this change does
not break old code.  Applications compiled with the old interface on such
machines will work the same way as they did before without recompiling.
Standard C compilers that support prototypes permit using int values for
dimension size and automatically convert them to longs before the call.  C
code that uses pointers to ints instead of pointers to longs (where required

mfhdf/hdp/hdp.c
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * Copyright by The HDF Group.                                               *
 * Copyright by the Board of Trustees of the University of Illinois.         *
 * All rights reserved.                                                      *
 *                                                                           *
 * This file is part of HDF.  The full HDF copyright notice, including       *
 * terms governing use, modification, and redistribution, is contained in    *
 * the COPYING file, which can be found at the root of the source code       *
 * distribution tree, or in https://support.hdfgroup.org/ftp/HDF/releases/.  *
 * If you do not have access to either file, you may request a copy from     *
 * help@hdfgroup.org.                                                        *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include "hdp.h"

#include "nc_priv.h"
#include "vg_priv.h"

/********************/
/* Global Variables */
/********************/

/* indicates Vsets have been initialized for the current file */
int vinit_done = FALSE;

/********************************/
/* Local variables and typedefs */
/********************************/

/* hdp commands (stored as (value, name) pairs to keep them in sync) */
typedef enum { HELP, LIST, DUMPSDS, DUMPRIG, DUMPVG, DUMPVD, DUMPGR, BAD_COMMAND } command_value_t;

typedef struct command_t {
    const command_value_t value;
    const char           *name;
} command_t;

static const command_t commands[] = {{HELP, "help"},       {LIST, "list"},
                                     {DUMPSDS, "dumpsds"}, {DUMPRIG, "dumprig"},
                                     {DUMPVG, "dumpvg"},   {DUMPVD, "dumpvd"},
                                     {DUMPGR, "dumpgr"},   {BAD_COMMAND, "BADNESS - not a valid command"}};

/* Print the usage message about this utility */
static void
usage(int argc, char *argv[])
{
    (void)argc;

    printf("%s, %s\n\n", argv[0], LIBVER_STRING);
    printf("Usage: hdp [-H] command [command options] <filelist>\n");
    printf("\t -H  Display usage information about the specified command.\n");
    printf("\t     If no command is specified, -H lists all commands.\n");
    printf("\t Commands:\n");
    printf("\t     list \tlists contents of files in <filelist>\n");
    printf("\t     dumpsds\tdisplays data of SDSs in <filelist>\n");
    printf("\t     dumpvd\tdisplays data of vdatas in <filelist>. \n");
    printf("\t     dumpvg\tdisplays data of vgroups in <filelist>. \n");
    printf("\t     dumprig\tdisplays data of RIs (DFR8 and DFR24) in <filelist>. \n");
    printf("\t     dumpgr\tdisplays data of RIs in <filelist>. \n");
    printf("\t <filelist>\tlist of hdf file names, separated by spaces.\n");
}

void
init_dump_opts(dump_info_t *dump_opts)
{
    dump_opts->filter             = DALL; /* default dump all GRs */
    dump_opts->by_index.num_list  = NULL; /* no index given */
    dump_opts->by_index.num_items = 0;
    dump_opts->by_ref.num_list    = NULL; /* no ref# given */
    dump_opts->by_ref.num_items   = 0;
    dump_opts->by_name.str_list   = NULL; /* no name given */
    dump_opts->by_name.num_items  = 0;
    dump_opts->by_class.str_list  = NULL; /* no class given */
    dump_opts->by_class.num_items = 0;
    dump_opts->contents           = DVERBOSE; /* default dump all information */
    dump_opts->dump_to_file       = FALSE;    /* don't dump to output file */
    dump_opts->file_format        = DASCII;   /* default output is ASCII file */
    dump_opts->file_type          = HDF_FILE; /* assuming an HDF file is provided */
    dump_opts->print_pal          = FALSE;    /* GR only, don't print palette */
    dump_opts->keep_order         = FALSE;
    dump_opts->all_types          = NULL;

    /* no specific dataset requested, default to dump all datasets */
    dump_opts->num_chosen = NO_SPECIFIC;

    /* print output aligned, using carriage returns */
    dump_opts->as_stream = FALSE;

    /* print space characters (LF, FF, CR, space, tabs...) in \digit format */
    dump_opts->clean_output = FALSE;

    /* print data starting at column 16 unless reset otherwise */
    dump_opts->firstln_indent = 16;

    /* print data on a continuous line starting at column 16 unless
       reset otherwise */
    dump_opts->contln_indent = 16;

    /* GR only, print data using interlace at creation */
    dump_opts->interlace = NO_SPECIFIC;

    /* GR & SD only, print data of global attributes unless -g is given */
    dump_opts->no_gattr_data = FALSE;

    /* GR & SD only, print data of local attributes unless -l is given */
    dump_opts->no_lattr_data = FALSE;

    strcpy(dump_opts->file_name, "\0");
} /* end init_dump_opts() */

int
main(int argc, char *argv[])
{
    command_value_t cmd = BAD_COMMAND; /* command to perform */
    int             curr_arg;          /* current cmd line argument */
    dump_opt_t      glob_opts;         /* global options for all commands */
    int             j;                 /* local counting variables */

    memset(&glob_opts, 0, sizeof(dump_opt_t));

    if (argc < 2) {
        usage(argc, argv);
        exit(EXIT_FAILURE);
    }

    curr_arg = 1;
    while (curr_arg < argc && (argv[curr_arg][0] == '-')) {
        switch (argv[curr_arg][1]) {
            case 'H':
                /* case 'h':  */ /*    Print help for a given command */
                if (curr_arg < argc - 1) {
                    glob_opts.help = TRUE; /* for displaying options. */
                    break;
                }
            default:
                usage(argc, argv); /* Display the general usage. */
                exit(EXIT_FAILURE);
        }
        curr_arg++;
    }

    for (j = 0; j < (sizeof(commands) / sizeof(command_t)); j++) {
        if (strcmp(argv[curr_arg], commands[j].name) == 0) {
            cmd = commands[j].value;
            break;
        }
    }
    if (cmd == BAD_COMMAND) {
        printf("Invalid command: %s\n", argv[curr_arg]);
        exit(EXIT_FAILURE);
    }

    curr_arg++;

    switch (cmd) {
        case LIST:
            if (FAIL == do_list(curr_arg, argc, argv, glob_opts.help))
                exit(EXIT_FAILURE);
            break;

        case DUMPSDS:
            if (FAIL == do_dumpsds(curr_arg, argc, argv, glob_opts.help))
                exit(EXIT_FAILURE);
            break;

        case DUMPRIG:
            /* BMR: retire dumprig, have dumpgr do the work */
            if (FAIL == do_dumprig(curr_arg, argc, argv, glob_opts.help))
                /*
                          fprintf( stderr, ">>> Please make a note that dumprig is no longer available.\n");
                          fprintf( stderr, "    The command dumpgr is and should be used in its place.\n" );
                          if (FAIL == do_dumpgr(curr_arg, argc, argv, glob_opts.help)) */
                exit(EXIT_FAILURE);
            break;

        case DUMPVG:
            if (FAIL == do_dumpvg(curr_arg, argc, argv, glob_opts.help))
                exit(EXIT_FAILURE);
            break;

        case DUMPVD:
            if (FAIL == do_dumpvd(curr_arg, argc, argv, glob_opts.help))
                exit(EXIT_FAILURE);
            break;

        case DUMPGR:
            if (FAIL == do_dumpgr(curr_arg, argc, argv, glob_opts.help))
                exit(EXIT_FAILURE);
            break;

        case HELP:
            usage(argc, argv);
            break;

        /* Should not get here - invalid commands are handled earlier */
        case BAD_COMMAND:
        default:
            printf("Invalid command!\n");
            exit(EXIT_FAILURE);
            break;
    }

    return EXIT_SUCCESS;
}

/* -----------------------------------------------------------------
NAME
   VShdfsize - computes the byte size of the field(s) of a vdata.

DESCRIPTION
   The size is the byte size of the fields of a vdata in a hdf file.
   This routine is very similar to the HDF API routine VSsizeof except
   it uses (struct vdata_desc).wlist.isize to compute the field size
   instead of (struct vdata_desc).wlist.esize as VSsizeof.

RETURNS
   The byte size of the field(s), positive integer, on success;
   otherwise, returns FAIL.
----------------------------------------------------------------- */
int32
VShdfsize(int32 vkey, /* IN vdata key */
          char *fields /* IN: Name(s) of the fields to check size of */)
{
    int32         totalsize;
    int32         i, j;
    int32         found;
    int32         ac;
    char        **av        = NULL;
    vsinstance_t *w         = NULL;
    VDATA        *vs        = NULL;
    int32         ret_value = SUCCEED;

    /* check key is valid vdata */
    if (HAatom_group(vkey) != VSIDGROUP)
        HGOTO_ERROR(DFE_ARGS, FAIL);

    /* get vdata instance */
    if (NULL == (w = (vsinstance_t *)HAatom_object(vkey)))
        HGOTO_ERROR(DFE_NOVS, FAIL);

    /* get vdata itself and check it */
    vs = w->vs;
    if (vs == NULL)
        HGOTO_ERROR(DFE_ARGS, FAIL);

    totalsize = 0;
    if (fields == NULL) /* default case? */
    {                   /* count all field sizes in vdata */
        for (j = 0; j < vs->wlist.n; j++)
            totalsize += vs->wlist.isize[j];
    }
    else if (fields[0] != '\0') /* implies: return 0 for empty 'fields' */
    {                           /* parse field string */
        if ((scanattrs(fields, &ac, &av) < 0) || (ac < 1))
            HGOTO_ERROR(DFE_ARGS, FAIL);

        for (i = 0; i < ac; i++) { /* check fields in vs */
            for (found = 0, j = 0; j < vs->wlist.n; j++)
                if (!strcmp(av[i], vs->wlist.name[j])) {
                    totalsize += vs->wlist.isize[j];
                    found = 1;
                    break;
                }

            if (!found)
                HGOTO_ERROR(DFE_ARGS, FAIL);
        } /* end for */
    }     /* end else */

    /* return total size of vdata fields specified */
    ret_value = totalsize;

done:
    return ret_value;
} /* VShdfsize */

/* ------------- VSattrhdfsize --------------------------
NAME
       VSattrhdfsize -- get hdfsize of a vdata attribute
USAGE
      int VSattrhdfsize(int32 vsid, int32 findex, int attrindex, int32 *size);
      int32 vsid;      IN: vdata id
      int32 findex;    IN: field index. _HDF_VDATA (-1) for the vdata
      int attrindex;  IN: which attr of the field/vdata
                           attrindex is 0-based
      int32 *size;     OUT: size of the attr values in hdf files.
RETURNS
        Returns SUCCEED when successful, FAIL otherwise.
DESCRIPTION
        size can be NULL if which is not interested.
--------------------------------------------------- */
int
VSattrhdfsize(int32 vsid, int32 findex, int attrindex, int32 *size)
{

    VDATA          *vs, *attr_vs;
    vs_attr_t      *vs_alist;
    vsinstance_t   *vs_inst, *attr_inst;
    int32           attr_vsid;
    int32           ret_value = SUCCEED;
    int             i, nattrs, a_index, found;
    DYN_VWRITELIST *w;

    HEclear();
    if (HAatom_group(vsid) != VSIDGROUP)
        HGOTO_ERROR(DFE_ARGS, FAIL);
    /* locate vs' index in vstab */
    if (NULL == (vs_inst = (vsinstance_t *)HAatom_object(vsid)))
        HGOTO_ERROR(DFE_NOVS, FAIL);
    if (NULL == (vs = vs_inst->vs))
        HGOTO_ERROR(DFE_NOVS, FAIL);
    if ((findex >= vs->wlist.n || findex < 0) && (findex != _HDF_VDATA))
        HGOTO_ERROR(DFE_BADFIELDS, FAIL);
    nattrs = vs->nattrs;
    if (attrindex < 0 || attrindex >= nattrs)
        HGOTO_ERROR(DFE_ARGS, FAIL);
    vs_alist = vs->alist;
    if (nattrs == 0 || vs_alist == NULL)
        /* no attrs or bad attr list */
        HGOTO_ERROR(DFE_ARGS, FAIL);
    found   = 0;
    a_index = -1;
    for (i = 0; i < nattrs; i++) {
        if (vs_alist->findex == findex) {
            a_index++;
            if (a_index == attrindex) {
                found = 1;
                break;
            }
        }
        vs_alist++;
    }
    if (!found)
        HGOTO_ERROR(DFE_ARGS, FAIL);
    /* found. get attr info */
    if (FAIL == (attr_vsid = VSattach(vs->f, (int32)vs_alist->aref, "r")))
        HGOTO_ERROR(DFE_CANTATTACH, FAIL);
    if (NULL == (attr_inst = (vsinstance_t *)HAatom_object(attr_vsid)))
        HGOTO_ERROR(DFE_NOVS, FAIL);
    if (NULL == (attr_vs = attr_inst->vs))
        HGOTO_ERROR(DFE_BADATTR, FAIL);

    /* If this vdata is not storing an attribute, it shouldn't have been passed
       to this function */
    if (strcmp(attr_vs->vsclass, _HDF_ATTRIBUTE) != 0)
        HGOTO_ERROR(DFE_BADATTR, FAIL);

    w = &(attr_vs->wlist);
    /* This vdata should have only 1 field */
    /* if (w->n != 1 || strcmp(w->name[0], ATTR_FIELD_NAME)) <- commented out
       Note: ATTR_FIELD_NAME cannot be used here because hdfeos sets fieldname
       to "AttrValues", not ATTR_FIELD_NAME. -BMR, 2014/12/01 */
    if (w->n != 1)
        HGOTO_ERROR(DFE_BADATTR, FAIL);
    if (size)
        *size = w->order[0] * (DFKNTsize(w->type[0]));
    if (FAIL == VSdetach(attr_vsid))
        HGOTO_ERROR(DFE_CANTDETACH, FAIL);
done:
    return ret_value;
} /* VSattrhdfsize */

/* ----------   Vattrhdfsize ----------------------
NAME
       Vattrhdfsize -- get hdfsize of a vgroup attribute
USAGE
        int Vattrhdfsize(int32 vgid, int attrindex, int32 *size)
        int32 vgid;      IN: vgroup id
        int attrindex;  IN: which attr's info we want
                             attrindex is 0-based
        int32 *size;     OUT: size of the attr values in hdf files.

RETURNS
        Returns SUCCEED when successful, FAIL otherwise.
DESCRIPTION
        size can be NULL if which is not interested.
--------------------------------------------------- */
int
Vattrhdfsize(int32 vgid, int attrindex, int32 *size)
{
    VGROUP         *vg;
    VDATA          *vs;
    DYN_VWRITELIST *w;
    vginstance_t   *v;
    vsinstance_t   *vs_inst;
    vg_attr_t      *vg_alist = NULL;
    int             adjusted_index;
    int32           fid, vsid;
    int32           ret_value = SUCCEED;

    HEclear();
    if (HAatom_group(vgid) != VGIDGROUP)
        HGOTO_ERROR(DFE_ARGS, FAIL);
    /* locate vg's index in vgtab */
    if (NULL == (v = (vginstance_t *)HAatom_object(vgid)))
        HGOTO_ERROR(DFE_VTAB, FAIL);
    vg  = v->vg;
    fid = vg->f;
    if (vg == NULL)
        HGOTO_ERROR(DFE_BADPTR, FAIL);
    if (vg->otag != DFTAG_VG)
        HGOTO_ERROR(DFE_ARGS, FAIL);

    /* Validate arguments */

    /* Check given index */
    if (attrindex < 0)
        HGOTO_ERROR(DFE_BADATTR, FAIL); /* invalid attribute index given */

    /* Check attribute list; it's complicated here to work around the old/new-style
       attribute issue, see headers of Vnattrs2, Vattrinfo2 in src for details */
    adjusted_index = attrindex;
    if (adjusted_index < vg->noldattrs) /* index of old-style attribute */
        vg_alist = vg->old_alist;       /* use old-attr list */
    else if (adjusted_index >= vg->noldattrs && adjusted_index < (vg->nattrs + vg->noldattrs))
    /* index of new-style attributes */
    {
        /* Adjust the index to accommodate for the old-style attributes
           preceding the new-style attribute list */
        adjusted_index = adjusted_index - vg->noldattrs;
        vg_alist       = vg->alist; /* use new-attr list */
    }
    else /* not that many attrs */
        HGOTO_ERROR(DFE_BADATTR, FAIL);

    if (vg_alist == NULL)
        /* Bad attr list */
        HGOTO_ERROR(DFE_BADATTR, FAIL);

    /* Get access to the vdata storing the attr, and obtain requested info */
    if ((vsid = VSattach(fid, (int32)vg_alist[attrindex].aref, "r")) == FAIL)
        HGOTO_ERROR(DFE_CANTATTACH, FAIL);
    if (HAatom_group(vsid) != VSIDGROUP)
        HGOTO_ERROR(DFE_ARGS, FAIL);
    if (NULL == (vs_inst = (vsinstance_t *)HAatom_object(vsid)))
        HGOTO_ERROR(DFE_NOVS, FAIL);
    if (NULL == (vs = vs_inst->vs))
        HGOTO_ERROR(DFE_BADATTR, FAIL);

    /* If this vdata is not storing an attribute, it shouldn't have been passed
       to this function */
    if (strcmp(vs->vsclass, _HDF_ATTRIBUTE) != 0)
        HGOTO_ERROR(DFE_BADATTR, FAIL);

    w = &(vs->wlist);
    /* This vdata should have only 1 field */
    /* if (w->n != 1 || strcmp(w->name[0], ATTR_FIELD_NAME)) <- commented out
       Note: ATTR_FIELD_NAME cannot be used here because hdfeos sets fieldname
       to "AttrValues", not ATTR_FIELD_NAME. -BMR, 2014/12/01 */
    if (w->n != 1)
        HGOTO_ERROR(DFE_BADATTR, FAIL);
    if (size)
        *size = w->order[0] * (DFKNTsize(w->type[0]));
    if (FAIL == VSdetach(vsid))
        HGOTO_ERROR(DFE_CANTDETACH, FAIL);
done:
    return ret_value;
} /* Vattrhdfsize */
